File: Python/fractionalization.py
# coding: utf-8
# © 2018 Aaron Sami Abassi
# Licensed under the Academic Free License version 3.0
FractionalType = type( 'Fractional', (), {} ) 
ArithmeticalType = type( 'Arithmetical', (), {} ) 
RelationalType = type( 'Relational', (), {} ) 
OperationalType = type( 'Operational', (), {} ) 
def Fractional( numerator = 0, denominator = 1 ):
    fraction = FractionalType()
    fraction.numerator = numerator
    fraction.denominator = denominator
    return fraction
def Arithmetical( add, subtract, multiply, divide ):
    arithmetic = ArithmeticalType()
    arithmetic.add = add
    arithmetic.subtract = subtract
    arithmetic.multiply = multiply
    arithmetic.divide = divide
    return arithmetic
def Relational( lesser, greater, equal, not_greater, not_lesser, not_equal ):
    relation = RelationalType()
    relation.lesser = lesser
    relation.greater = greater
    relation.equal = equal
    relation.not_greater = not_greater
    relation.not_lesser = not_lesser
    relation.not_equal = not_equal
    return relation
def Operational( arithmetic, relation ):
    operation = OperationalType()
    operation.arithmetic = arithmetic
    operation.relation = relation
    return operation
def greatest_common_divisor( a, b ):
    # Euclidean Algorithm
    if a > b:
        dividend = a
        divisor = b
    else:
        dividend = b
        divisor = a
    remainder = dividend % divisor
    while remainder > 0:
        dividend = divisor
        divisor = remainder
        remainder = dividend % divisor
    return divisor
def least_common_multiple( a, b ):
    divisor = greatest_common_divisor( a, b )
    if a > b:
        return int( a / divisor * b )
    return int( b / divisor * a )
def reduce( fraction ):
    if fraction.numerator == 0:
        result = Fractional()
    else:
        divisor = greatest_common_divisor( fraction.numerator, fraction.denominator )
        result = Fractional( int(fraction.numerator / divisor), int(fraction.denominator / divisor) )
    return result
def fast_add( base, relative ):
    return Fractional( 
        base.numerator * relative.denominator + relative.numerator * base.denominator,
        base.denominator * relative.denominator
    )
def fast_subtract( base, relative ):
    return Fractional(
        base.numerator * relative.denominator - relative.numerator * base.denominator,
        base.denominator * relative.denominator
    )
def fast_multiply( base, relative ):
    return Fractional(
        base.numerator * relative.numerator,
        base.denominator * relative.denominator
    )
def fast_divide( base, relative ):
    return Fractional(
        base.numerator * relative.denominator,
        base.denominator * relative.numerator
    )
def reducing_add( base, relative ):
    result = Fractional()
    result.denominator = least_common_multiple( base.denominator, relative.denominator )
    result.numerator = base.numerator * int(result.denominator / base.denominator)
    result.numerator += relative.numerator * int(result.denominator / relative.denominator)
    return reduce( result )
def reducing_subtract( base, relative ):
    result = Fractional()
    result.denominator = least_common_multiple( base.denominator, relative.denominator )
    result.numerator = base.numerator * int(result.denominator / base.denominator)
    result.numerator -= relative.numerator * int(result.denominator / relative.denominator)
    return reduce( result )
def reducing_multiply( base, relative ):
    if base.numerator > 0:
        bnrd_divisor = greatest_common_divisor( base.numerator, relative.denominator )
    else:
        bnrd_divisor = 1
    if relative.numerator > 0:
        bdrn_divisor = greatest_common_divisor( base.denominator, relative.numerator )
    else:
        bdrn_divisor = 1
    result = Fractional()
    result.numerator = int(base.numerator / bnrd_divisor) * int(relative.numerator / bdrn_divisor )
    result.denominator = int(base.denominator / bdrn_divisor) * int(relative.denominator / bnrd_divisor )
    return result
def reducing_divide( base, relative ):
    bnrn_divisor = greatest_common_divisor( base.numerator, relative.numerator )
    bdrd_divisor = greatest_common_divisor( base.denominator, relative.denominator )
    result = Fractional()
    result.numerator = int(base.numerator / bnrn_divisor) * int(relative.denominator / bdrd_divisor )
    result.denominator = int(base.denominator / bdrd_divisor) * int(relative.numerator / bnrn_divisor )
    return result
def fast_lesser( base, relative ):
    return base.numerator * relative.denominator < relative.numerator * base.denominator
def fast_greater( base, relative ):
    return base.numerator * relative.denominator > relative.numerator * base.denominator
def fast_equal( base, relative ):
    return base.numerator * relative.denominator == relative.numerator * base.denominator
def fast_not_greater( base, relative ):
    return base.numerator * relative.denominator <= relative.numerator * base.denominator
def fast_not_lesser( base, relative ):
    return base.numerator * relative.denominator >= relative.numerator * base.denominator
def fast_not_equal( base, relative ):
    return base.numerator * relative.denominator != relative.numerator * base.denominator
def reducing_lesser( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) < relative.numerator * int(factor / relative.denominator)
def reducing_greater( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) > relative.numerator * int(factor / relative.denominator)
def reducing_equal( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) == relative.numerator * int(factor / relative.denominator)
def reducing_not_greater( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) <= relative.numerator * int(factor / relative.denominator)
def reducing_not_lesser( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) >= relative.numerator * int(factor / relative.denominator)
def reducing_not_equal( base, relative ):
    factor = least_common_multiple( base.denominator, relative.denominator )
    return base.numerator * int(factor / base.denominator) != relative.numerator * int(factor / relative.denominator)
fast_arithmetic = Arithmetical( fast_add, fast_subtract, fast_multiply, fast_divide )
reducing_arithmetic = Arithmetical( reducing_add, reducing_subtract, reducing_multiply, reducing_divide )
fast_relation = Relational( fast_lesser, fast_greater, fast_equal, fast_not_greater, fast_not_lesser, fast_not_equal )
reducing_relation = Relational( reducing_lesser, reducing_greater, reducing_equal, reducing_not_greater, reducing_not_lesser, reducing_not_equal )
fast_operation = Operational( fast_arithmetic, fast_relation )
reducing_operation = Operational( reducing_arithmetic, reducing_relation )

File: VB/Fractionalization.vb
' © 2018 Aaron Sami Abassi
' Licensed under the Academic Free License version 3.0
Module Fractionalization

    Public Delegate Function ArithmeticalAbstract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional

    Public Delegate Function RelationalAbstract(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean

    Public Structure Fractional

        Dim Numerator, Denominator As UInteger

        Sub New(ByRef Numerator As UInteger, ByRef Denominator As UInteger)
            Me.Numerator = Numerator
            Me.Denominator = Denominator
        End Sub

    End Structure

    Public Structure Arithmetical

        ReadOnly Add, Subtract, Multiply, Divide As ArithmeticalAbstract

        Sub New(ByRef Add As ArithmeticalAbstract, ByRef Subtract As ArithmeticalAbstract, ByRef Multiply As ArithmeticalAbstract, ByRef Divide As ArithmeticalAbstract)
            Me.Add = Add
            Me.Subtract = Subtract
            Me.Multiply = Multiply
            Me.Divide = Divide
        End Sub

    End Structure

    Public Structure Relational

        ReadOnly Lesser, Greater, Equal, NotGreater, NotLesser, NotEqual As RelationalAbstract

        Sub New(ByRef Lesser As RelationalAbstract, ByRef Greater As RelationalAbstract, ByRef Equal As RelationalAbstract, ByRef NotGreater As RelationalAbstract, ByRef NotLesser As RelationalAbstract, ByRef NotEqual As RelationalAbstract)
            Me.Lesser = Lesser
            Me.Greater = Greater
            Me.Equal = Equal
            Me.NotGreater = NotGreater
            Me.NotLesser = NotLesser
            Me.NotEqual = NotEqual
        End Sub

    End Structure

    Public Structure Operational

        ReadOnly Arithmetic As Arithmetical

        ReadOnly Relation As Relational

        Sub New(ByRef Arithmetic As Arithmetical, ByRef Relation As Relational)
            Me.Arithmetic = Arithmetic
            Me.Relation = Relation
        End Sub

    End Structure

    Public Function GreatestCommonDivisor(A As UInteger, B As UInteger) As UInteger
        ' Eclidean Algorithm
        Dim Dividend, Divisor, Remainder As UInteger
        If A > B Then
            Dividend = A
            Divisor = B
        Else
            Dividend = B
            Divisor = A
        End If
        While (Remainder = Dividend Mod Divisor) > 0
            Dividend = Divisor
            Divisor = Remainder
        End While
        Return Divisor
    End Function

    Public Function LeastCommonMultiple(A As UInteger, B As UInteger) As UInteger
        Dim Divisor As UInteger
        Divisor = GreatestCommonDivisor(A, B)
        If A > B Then
            Return A / Divisor * B
        End If
        Return B / Divisor * A
    End Function

    Public Function Reduce(ByRef Fraction As Fractional) As Fractional
        Dim Divisor As UInteger
        Dim Result As Fractional
        If Fraction.Numerator = 0 Then
            Result.Numerator = 0
            Result.Denominator = 1
        Else
            Divisor = GreatestCommonDivisor(Fraction.Numerator, Fraction.Denominator)
            Result.Numerator = Fraction.Numerator / Divisor
            Result.Denominator = Fraction.Denominator / Divisor
        End If
        Return Result
    End Function

    Public Function FastAdd(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator + Relative.Numerator * Base.Denominator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastSubtract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator - Relative.Numerator * Base.Denominator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastMultiply(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Numerator
        Result.Denominator = Base.Denominator * Relative.Denominator
        Return Result
    End Function

    Public Function FastDivide(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Numerator = Base.Numerator * Relative.Denominator
        Result.Denominator = Base.Denominator * Relative.Numerator
        Return Result
    End Function

    Public Function ReducingAdd(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Denominator = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Result.Numerator = Base.Numerator * (Result.Denominator / Base.Denominator)
        Result.Numerator += Relative.Numerator * (Result.Denominator / Relative.Denominator)
        Return Reduce(Result)
    End Function

    Public Function ReducingSubtract(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim Result As Fractional
        Result.Denominator = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Result.Numerator = Base.Numerator * (Result.Denominator / Base.Denominator)
        Result.Numerator -= Relative.Numerator * (Result.Denominator / Relative.Denominator)
        Return Reduce(Result)
    End Function

    Public Function ReducingMultiply(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim LNRDDivisor, LDRNDivisor As UInteger
        Dim Result As Fractional
        LNRDDivisor = GreatestCommonDivisor(Base.Numerator, Relative.Denominator)
        LDRNDivisor = GreatestCommonDivisor(Base.Denominator, Relative.Numerator)
        Result.Numerator = (Base.Numerator / LNRDDivisor) * (Relative.Numerator / LDRNDivisor)
        Result.Denominator = (Base.Denominator / LDRNDivisor) * (Relative.Denominator / LNRDDivisor)
        Return Result
    End Function

    Public Function ReducingDivide(ByRef Base As Fractional, ByRef Relative As Fractional) As Fractional
        Dim LNRNDivisor, LDRDDivisor As UInteger
        Dim Result As Fractional
        LNRNDivisor = GreatestCommonDivisor(Base.Numerator, Relative.Numerator)
        LDRDDivisor = GreatestCommonDivisor(Base.Denominator, Relative.Denominator)
        Result.Numerator = (Base.Numerator / LNRNDivisor) * (Relative.Denominator / LDRDDivisor)
        Result.Denominator = (Base.Denominator / LDRDDivisor) * (Relative.Numerator / LNRNDivisor)
        Return Result
    End Function

    Public Function FastLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator < Relative.Numerator * Base.Denominator
    End Function

    Public Function FastGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator > Relative.Numerator * Base.Denominator
    End Function

    Public Function FastEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator = Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator <= Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator >= Relative.Numerator * Base.Denominator
    End Function

    Public Function FastNotEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Return Base.Numerator * Relative.Denominator <> Relative.Numerator * Base.Denominator
    End Function

    Public Function ReducingLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) < Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) > Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) = Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotGreater(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) <= Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotLesser(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) >= Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public Function ReducingNotEqual(ByRef Base As Fractional, ByRef Relative As Fractional) As Boolean
        Dim Factor As UInteger = LeastCommonMultiple(Base.Denominator, Relative.Denominator)
        Return Base.Numerator * (Factor / Base.Denominator) <> Relative.Numerator * (Factor / Relative.Denominator)
    End Function

    Public ReadOnly FastArithmetic As New Arithmetical(AddressOf FastAdd, AddressOf FastSubtract, AddressOf FastMultiply, AddressOf FastDivide)

    Public ReadOnly ReducingArithmetic As New Arithmetical(AddressOf ReducingAdd, AddressOf ReducingSubtract, AddressOf ReducingMultiply, AddressOf ReducingDivide)

    Public ReadOnly FastRelation As New Relational(AddressOf FastLesser, AddressOf FastGreater, AddressOf FastEqual, AddressOf FastNotGreater, AddressOf FastNotLesser, AddressOf FastNotEqual)

    Public ReadOnly ReducingRelation As New Relational(AddressOf ReducingLesser, AddressOf ReducingGreater, AddressOf ReducingEqual, AddressOf ReducingNotGreater, AddressOf ReducingNotLesser, AddressOf ReducingNotEqual)

    Public ReadOnly FastOperation As New Operational(FastArithmetic, FastRelation)

    Public ReadOnly ReducingOperation As New Operational(ReducingArithmetic, ReducingRelation)

End Module

Module Example

    Sub DisplayFraction(ByRef Fraction As Fractional)
        Console.Write(Fraction.Numerator)
        If (Fraction.Denominator <> 1) Then
            Console.Write("/" & Fraction.Denominator)
        End If
    End Sub

    Sub DisplayArithmetic(ByRef Base As Fractional, Symbol As String, ByRef Relative As Fractional, ByRef Equals As Fractional)
        DisplayFraction(Base)
        Console.Write(" " & Symbol & " ")
        DisplayFraction(Relative)
        Console.Write(" = ")
        DisplayFraction(Equals)
        Console.WriteLine()
    End Sub

    Sub DisplayRelation(ByRef Base As Fractional, Symbol As String, ByRef Relative As Fractional, ByRef Result As Boolean)
        DisplayFraction(Base)
        Console.Write(" " & Symbol & " ")
        DisplayFraction(Relative)
        Console.WriteLine(" = " & Result)
    End Sub

    Sub DisplayOperations(ByRef Operation As Operational, ByVal Base As Fractional, ByVal Relative As Fractional)
        Dim Arithmetic As Arithmetical = Operation.Arithmetic
        Dim Relation As Relational = Operation.Relation
        DisplayArithmetic(Base, "+", Relative, Arithmetic.Add(Base, Relative))
        DisplayArithmetic(Base, "-", Relative, Arithmetic.Subtract(Base, Relative))
        DisplayArithmetic(Base, "*", Relative, Arithmetic.Multiply(Base, Relative))
        DisplayArithmetic(Base, "/", Relative, Arithmetic.Divide(Base, Relative))
        DisplayRelation(Base, "<", Relative, Relation.Lesser(Base, Relative))
        DisplayRelation(Base, ">", Relative, Relation.Greater(Base, Relative))
        DisplayRelation(Base, "=", Relative, Relation.Equal(Base, Relative))
        DisplayRelation(Base, "<=", Relative, Relation.NotGreater(Base, Relative))
        DisplayRelation(Base, ">=", Relative, Relation.NotLesser(Base, Relative))
        DisplayRelation(Base, "<>", Relative, Relation.NotEqual(Base, Relative))
    End Sub

    Sub Main()
        Static X As New Fractional(1, 6)
        Static Y As New Fractional(1, 12)
        Console.WriteLine("Fast Fractional Operations")
        DisplayOperations(FastOperation, X, Y)
        Console.WriteLine()
        Console.WriteLine("Reducing Fractional Operations")
        DisplayOperations(ReducingOperation, X, Y)
    End Sub

End Module
